### PR TITLE
Add remote snapshotter architecture doc

### PR DESCRIPTION
**Pull request by Kern--**
Signed-off-by: Kern Walster <walster@amazon.com>

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/remote-snapshotter.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+This document is about the architecture used to support remote snapshotters in firecracker-containerd. If you would like to jump in and launch estargz images on firecracker-containerd, see the [remote snapshotter getting started guide.](remote-snapshotter-getting-started.md)
```
Comment (ginglis13) [inline docs/remote-snapshotter.md:1]: nit: eStargz

File: docs/remote-snapshotter.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+There are several non-core containerd remote snapshotter projects which implement lazy loading via various mechanisms.
+* [stargz-snapshotter](https://github.com/containerd/stargz-snapshotter) uses a custom [estargz image format](https://github.com/containerd/stargz-snapshotter/blob/v0.12.0/docs/estargz.md) and exposes the lazily loaded image as a FUSE filesystem
```
Comment (ginglis13) [inline docs/remote-snapshotter.md:8]: nit: eStargz

File: docs/remote-snapshotter.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+* __credential-helper-mmds__ - A [Docker credential helper](https://github.com/docker/docker-credential-helpers) that can read credentials from MMDS. Remote snapshotters that support the Docker credential helper pattern can use this tool to retrieve the credentials needed to pull from the remote registry.
+* __socat__ - [A multipurpose relay](http://www.dest-unreach.org/socat/). It is used to expose unix sockets inside the microVM as vsock services that can be reached from the host.
+* __remote snapshotter__ - The snapshotter that is responsible for lazy loading container images
+
+## New components
+
+At a high level, this architecture adds 4 components:
+1) A remote snapshotter inside the Firecracker rootfs that launches at boot 
+1) Socat to expose the remote snapshotter's socket over vsock
+1) A [demux-snapshotter](https://github.com/firecracker-microvm/firecracker-containerd/blob/913ab8bfc13d4735f3cd62dc6ac5f65e80db4400/snapshotter/README.md) which exposes itself a single snapshotter to containerd, but proxies snapshot requests to the in-VM remote snapshotters as they come and go via vsock
```
Comment (ginglis13) [inline docs/remote-snapshotter.md:40]: 1/ itself as a

2/ nit: I'm in favor of omitting "as they come and go" to make this "to the in-VM remote snapshotters via vsock"

File: docs/remote-snapshotter.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+1) [docker-credential-helper-mmds](https://github.com/firecracker-microvm/firecracker-containerd/blob/913ab8bfc13d4735f3cd62dc6ac5f65e80db4400/docker-credential-mmds) to expose Firecracker's MMDS via the credential helper interface
+
+## Credentials
+
+Each remote snapshotter has its own mechanism for retrieving registry credentials. Many involve reading directly from a file on disk which is not very user-friendly in the firecracker-containerd case where that file would either need to be be baked into the readonly rootfs or attached to the microVM as a block device.
```
Comment (austinvazquez) [inline docs/remote-snapshotter.md:45]: to be be baked -> to be baked

**Comment by austinvazquez** [comment]
Diagrams look great!

File: docs/img/remote-snapshotter-launch-sequence.plantuml [hunk 5]
```diff
@@ -0,0 +1,66 @@
+@startuml
```
Comment (kzys) [inline docs/img/remote-snapshotter-launch-sequence.plantuml:1]: This is pretty cool!

File: docs/remote-snapshotter.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+# Remote snapshotters overview
+
+Remote snapshotters are a type of [containerd snapshotter](https://github.com/containerd/containerd/blob/ff91434af1b937a31cd3f76c31d079b74a11d9d0/docs/snapshotters/README.md) that are responsible for fetching snapshots from a remote location. When pulling an image with a normal snapshotter, the snapshotter provides mount points, but the containerd client pulls image layers into the containerd content store and unpacks them onto those mount points. With remote snapshotters, the snapshotter is responsible for providing mount points as well as pulling image layers and populating that snapshot. This allows remote snapshotters such as the [stargz-snapshotter](https://github.com/containerd/stargz-snapshotter) to lazily load snapshot content on-demand, thus reducing the upfront time needed for image pull. For more details see [containerd's remote snapshotter documentation](https://github.com/containerd/containerd/blob/v1.6.6/docs/remote-snapshotter.md).
```
Comment (kzys) [inline docs/remote-snapshotter.md:5]: How about using one of the release tags instead of raw Git commit hashes?

**Review by kzys — APPROVED** [review]

File: docs/remote-snapshotter.md (hunk 6 above)
Comment (Kern--) [reply docs/remote-snapshotter.md:5]: I tried to where I could, but the snapshotters readme was created after the 1.6 tag and won't be backported, so it's not in any tag until 1.7.

Comment (kzys) [reply docs/remote-snapshotter.md:5]: I didn't know that. Thanks!

**Review by ginglis13 — APPROVED** [review]

File: docs/remote-snapshotter.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+## Metrics
+
+For remote snapshotters that expose metrics via [Prometheus](https://prometheus.io/), the demux-snapshotter provides APIs for discovering and scraping each remote snapshotter running inside the microVM. 
+
+The demux-snapshotter provides 2 components for this:
+1) A Prometheus [http service discovery API](https://prometheus.io/) that returns an http endpoint for each remote snapshotter that can be scraped to pull that snapshotter's metrics
+2) An http proxy service for each remote snapshotter to proxy the HTTP request over the microVM's vsock to the remote snapshotter.
+
+For more deatils, see the [metrics configuration for the demux-snapshotter](https://github.com/firecracker-microvm/firecracker-containerd/tree/main/snapshotter#metrics).
```
Comment (austinvazquez) [inline docs/remote-snapshotter.md:59]: details* :)

**Review by austinvazquez — APPROVED** [review]

**Pull request merged**